### PR TITLE
rebasing moderner-cmake on gk

### PR DESCRIPTION
**Pull request by mathiaswagner**
I rebased #1120 on generic kernels branch.

I think this mostly works now.
Some things to look at:
* issues with QMP/QIO and them being build as static libraries by default (need to look into that)
* warnings from QIO/QMP are not suppressed
* USQCD is not readily available for MILC but you need to run `install` (which can now go into build_dir/usqcd)

We can iterate on these.

Quick reminder:

- [x]  bump QIO and QMP versions

### CODE REVIEW

File: CMakeLists.txt
Comment (bjoo) [inline CMakeLists.txt:481]: Should we check that QUDA_NVSMEM is on ?

File: CMakeLists.txt
Comment (bjoo) [inline CMakeLists.txt:512]: missing 

message(STATUS 
?

File: CMakeLists.txt
Comment (bjoo) [inline CMakeLists.txt]: Since this is quarded by QUDA_OPENMP  should we make it `REQUIRED` ?

File: CMakeLists.txt
Comment (bjoo) [inline CMakeLists.txt]: Since this is guarded by `QUDA_MPI` should we make this `REQUIRED` ?  `QMP` should bring it transitively so perhaps we can soften this? But otherwise it is fine...

File: CMakeLists.txt
Comment (bjoo) [inline CMakeLists.txt:651]: I CMakeified QDP_JIT so this will get simpler soon. I just need to see if it still attempts to hijack your allocator or not. It should transitively bring QMP, QIO etc.

File: QUDAConfig.cmake.in
Comment (bjoo) [inline QUDAConfig.cmake.in]: Should it be `REQUIRED`?

File: lib/CMakeLists.txt
Comment (bjoo) [inline lib/CMakeLists.txt]: Good practice to:

add_library( QUDA::quda ALIAS quda )

and have the tests link to QUDA::quda ?

File: tests/CMakeLists.txt
Comment (bjoo) [inline tests/CMakeLists.txt:15]: Should NVSHMEM be a dependency of the quda library to make it self contained?

File: QUDAConfig.cmake.in
Comment (bjoo) [inline QUDAConfig.cmake.in:35]: Should we add a SHMEM dependency here?

**Review by bjoo — APPROVED** [review]
This looks really nice. I have a few potential changes I made via comments. Certainly a good place to start from. If @maddyscientist and @mathiaswagner are both happy with it, I'd go with it. I can merge in HIP changes too in my branch. Working from this would be better than me redoing it all :). Nice work Mathias!

File: QUDAConfig.cmake.in
Comment (mathiaswagner) [reply QUDAConfig.cmake.in:35]: No, we don't need that. nvshmem is consumed into libquda.

File: tests/CMakeLists.txt
Comment (mathiaswagner) [reply tests/CMakeLists.txt:15]: Until we build nvshmem with cmake itself this is a hacky workaround we still need (I believe).

File: CMakeLists.txt
Comment (mathiaswagner) [reply CMakeLists.txt:651]: Cool. Happy to see that also getting cmakified.

File: CMakeLists.txt
Comment (mathiaswagner) [reply CMakeLists.txt]: Done.

**Comment by maddyscientist** [comment]
This looks like great progress.  
* My only real concern is around MILC not being able to use the USQCD libraries without installing.  While this may have an easy solution, `make install`, it will lead to a lot of increased disk space utilization for developers (speaking as someone who has to regularly watch their disk utilization given the number of users of nvsocal2 😉 ).  Is there a possible solution not involving `make install`, or is the only other solution to install qmp/qio outside of QUDA?  Not a blocker, but a definite nice to have.
* Looks like two of the Jenkins failures are due to insufficient cmake version, but the other failure looks a bit less trivial

File: CMakeLists.txt
Comment (mathiaswagner) [inline CMakeLists.txt:598]: @maddyscientist this sets the default install paths.

**Comment by mathiaswagner** [comment]
The issue with USQCD is that QIO/QMP are only usable after the make install step. Otherwise headers are scattered all over the source / build directories. If you are really worried about disk space you are probably fine with installing QIO/QMP on your machine once and just use that (not have QUDA build it)???

**Comment by maddyscientist** [comment]
> The issue with USQCD is that QIO/QMP are only usable after the make install step. Otherwise headers are scattered all over the source / build directories. If you are really worried about disk space you are probably fine with installing QIO/QMP on your machine once and just use that (not have QUDA build it)???

Yep, a separate install is ok.  As I said, it's not a blocker, just a nice-to-have feature

**Comment by mathiaswagner** [comment]
My main concern was that we don't need to teach others how to build QIO/QMP for MILC. But maybe I have another workaround for you 🤔

File: lib/CMakeLists.txt
Comment (maddyscientist) [inline lib/CMakeLists.txt]: indent has been screwed up here

**Review by maddyscientist — COMMENTED** [review]
Certainly passes visual review, beyond the trivial formatting comment.  Will test this in action now.

**Comment by maddyscientist** [comment]
Actual testing is revealing some issues:
* When enabling `QMP`, built with QUDA, we are subjected to a slew of warnings regarding doxygen.  This is unexpected since I see the option `"QMP_BUILD_DOCS" "OFF"` is set in CMakeList.txt 
```
warning: Tag 'COLS_IN_ALPHA_INDEX' at line 68 of file '/scratch/kate/quda-develop/_deps/qmp-build/doc/QMPdoxyfile' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'HTML_ALIGN_MEMBERS' at line 78 of file '/scratch/kate/quda-develop/_deps/qmp-build/doc/QMPdoxyfile' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'PERL_PATH' at line 132 of file '/scratch/kate/quda-develop/_deps/qmp-build/doc/QMPdoxyfile' has become obsolete.
```
* Similarly, there are a slew of warnings associated with building the QMP examples, even though the option looks to be disabled in CMakeLists.txt.  As I type this, @bjoo tells me he's going to fix these 😄

**Comment by maddyscientist** [comment]
Another issue: it seems that `Eigen` is not being included with `-isystem` and instead is being included with `-I`.  This leads to warnings with some compilers.  For example, `nvc++`
```
cd /scratch/kate/quda-develop-nvc++/lib && /opt/nvidia/hpc_sdk/Linux_x86_64/dev/compilers/bin/nvc++ -DBUILD_QDP_INTERFACE -DDEVEL -DEIGEN -DGITVERSION=\"PERFLAB_MILC_201903-3563-g815ac4763-dirty-sm_70\" -DGPU_WILSON_DIRAC -DNATIVE_LAPACK_LIB -DQUDA_FAST_COMPILE_DSLASH -DQUDA_FAST_COMPILE_REDUCE -DQUDA_HASH=\"cpu_arch=x86_64,gpu_arch=sm_70,cuda_version=99.99.205935\" -DQUDA_PRECISION=4 -DQUDA_RECONSTRUCT=4 -I/scratch/kate/quda-develop-nvc++/lib -I/home/kate/github/quda-develop/lib -I/home/kate/github/quda-develop/lib/. -I/home/kate/github/quda-develop/lib/../include -I/scratch/kate/quda-develop-nvc++/include -I/home/kate/github/quda-develop/lib/../include/targets/cuda -I/scratch/kate/quda-develop-nvc++/include/targets/cuda -I/home/kate/github/quda-develop/lib/../include/targets/cuda/externals -I/home/kate/github/quda-develop/lib/../include/targets/generic -I/home/kate/github/quda-develop/lib/../include/externals -I/scratch/kate/quda-develop-nvc++/_deps/eigen-src -I/opt/nvidia/hpc_sdk/Linux_x86_64/dev/cuda/11.4/include -g -O3 -Wall -Wextra -fPIC --c++17 -o CMakeFiles/quda_cpp.dir/eigensolve_quda.cpp.o -c /home/kate/github/quda-develop/lib/eigensolve_quda.cpp
"/scratch/kate/quda-develop-nvc++/_deps/eigen-src/Eigen/src/Core/util/Memory.h", line 308: warning: statement is unreachable
    return NULL;
    ^
          detected during:
            instantiation of "T *Eigen::internal::conditional_aligned_realloc_new_auto<T,Align>(T *, std::size_t, std::size_t) [with T=std::complex<double>, Align=true]" at line 476 of "/scratch/kate/quda-develop-nvc++/_deps/eigen-src/Eigen/src/Core/DenseStorage.h"
            instantiation of "void Eigen::DenseStorage<T, -1, -1, -1, _Options>::conservativeResize(Eigen::Index, Eigen::Index, Eigen::Index) [with T=std::complex<double>, _Options=0]" at line 1017 of "/scratch/kate/quda-develop-nvc++/_deps/eigen-src/Eigen/src/Core/PlainObjectBase.h"
            instantiation of "void Eigen::internal::conservative_resize_like_impl<Derived, OtherDerived, IsVector>::run(Eigen::DenseBase<Derived> &, Eigen::Index, Eigen::Index) [with Derived=Eigen::Matrix<std::complex<double>, -1, -1, 0, -1, -1>, OtherDerived=Eigen::Matrix<std::complex<double>, -1, -1, 0, -1, -1>, IsVector=false]" at line 384 of "/scratch/kate/quda-develop-nvc++/_deps/eigen-src/Eigen/src/Core/PlainObjectBase.h"
            instantiation of "void Eigen::PlainObjectBase<Derived>::conservativeResize(Eigen::Index, Eigen::Index) [with Derived=Eigen::Matrix<std::complex<double>, -1, -1, 0, -1, -1>]" at line 983 of "/home/kate/github/quda-develop/lib/eigensolve_quda.cpp"
```
Looking at CMakeLists.txt, this is unexpected since `SYSTEM` is declared
```
target_include_directories(Eigen SYSTEM INTERFACE ${Eigen_SOURCE_DIR})
```

**Comment by bjoo** [comment]
As regards the Doxygen and test building they can be resolved with Judicious use of (") marks in the CMake LIsts
ie "OPTION VALUE" as opposed to "OPTION" "VALUE"
(as discussed with @maddyscientist on Slack)
Dunno about the Eigen stuff

**Comment by maddyscientist** [comment]
Ok, I have pushed fixes for `"QMP_BUILD_DOCS OFF"` and `"QMP_TESTING"` as well as updated the git hash for QMP to correspond to the latest version of QMP which fixes all the warnings on g++.  There are some outstanding warnings when `nvc++` is used to build QMP, but that shouldn't be blocking on this PR.  I can work on addressing those in a PR to QMP separate from this.

One funny thing I noticed is that when building QUDA with `nvc++` with QMP/QIO enabled, final linking failed unless I set the cpp compiler to be `mpicxx` and not `nvc++`.  Strangely this does not happen with just MPI enablement.
e.g.
```
cd /scratch/kate/quda-develop-nvc++/tests && /opt/cmake-nvhpc/bin/cmake -E cmake_link_script CMakeFiles/dslash_test.dir/link.txt --verbose=1
/opt/nvidia/hpc_sdk/Linux_x86_64/dev/compilers/bin/nvc++ -cuda -Wl,-rpath "-Wl,\$$ORIGIN:\$$ORIGIN/../../lib:\$$ORIGIN/../../../lib:\$$ORIGIN/../../../compilers/lib:\$$ORIGIN/../../../../compilers/lib:\$$ORIGIN/../../../../../compilers/lib" -Wl,-rpath -Wl,/opt/nvidia/hpc_sdk/Linux_x86_64/dev/comm_libs/openmpi/openmpi-3.1.5/lib @CMakeFiles/dslash_test.dir/objects1.rsp -o dslash_test  -Wl,-rpath,/scratch/kate/quda-develop-nvc++/lib:/scratch/kate/quda-develop-nvc++/_deps/qio-build/lib:/scratch/kate/quda-develop-nvc++/_deps/qmp-build/lib:/opt/nvidia/hpc_sdk/Linux_x86_64/dev/comm_libs/openmpi/openmpi-3.1.5/lib:/scratch/kate/quda-develop-nvc++/_deps/qio-build/other_libs/c-lime/lib: libquda_test.a ../lib/libquda.so /opt/nvidia/hpc_sdk/Linux_x86_64/dev/cuda/11.4/lib64/libcudart_static.a -ldl /usr/lib/x86_64-linux-gnu/librt.so /usr/lib/x86_64-linux-gnu/libcuda.so /usr/lib/x86_64-linux-gnu/libcublas.so ../_deps/qio-build/lib/libqio.so ../_deps/qmp-build/lib/libqmp.so /opt/nvidia/hpc_sdk/Linux_x86_64/dev/comm_libs/openmpi/openmpi-3.1.5/lib/libmpi.so ../_deps/qio-build/other_libs/c-lime/lib/liblime.so -ldl -lpthread 
/usr/bin/ld: warning: /opt/nvidia/hpc_sdk/Linux_x86_64/dev/compilers/lib/nvhpc.ld contains output sections; did you forget -T?
/usr/bin/ld: libquda_test.a(host_utils.cpp.o): in function `MPI::Op::Init(void (*)(void const*, void*, int, MPI::Datatype const&), bool)':
/opt/nvidia/hpc_sdk/Linux_x86_64/dev/comm_libs/mpi/include/openmpi/ompi/mpi/cxx/op_inln.h:121: undefined reference to `ompi_mpi_cxx_op_intercept'
/usr/bin/ld: libquda_test.a(host_utils.cpp.o): in function `MPI::Intracomm::Clone() const':
/opt/nvidia/hpc_sdk/Linux_x86_64/dev/comm_libs/mpi/include/openmpi/ompi/mpi/cxx/intracomm_inln.h:23: undefined reference to `MPI::Comm::Comm()'
/usr/bin/ld: libquda_test.a(host_utils.cpp.o): in function `MPI::Intracomm::Create(MPI::Group const&) const':
/opt/nvidia/hpc_sdk/Linux_x86_64/dev/comm_libs/mpi/include/openmpi/ompi/mpi/cxx/intracomm_inln.h:23: undefined reference to `MPI::Comm::Comm()'
/usr/bin/ld: libquda_test.a(host_utils.cpp.o): in function `MPI::Intracomm::Split(int, int) const':
/opt/nvidia/hpc_sdk/Linux_x86_64/dev/comm_libs/mpi/include/openmpi/ompi/mpi/cxx/intracomm_inln.h:23: undefined reference to `MPI::Comm::Comm()'
/usr/bin/ld: libquda_test.a(host_utils.cpp.o): in function `MPI::Intracomm::Create_cart(int, int const*, bool const*, bool) const':
/opt/nvidia/hpc_sdk/Linux_x86_64/dev/comm_libs/mpi/include/openmpi/ompi/mpi/cxx/intracomm_inln.h:28: undefined reference to `MPI::Comm::Comm()'
/usr/bin/ld: libquda_test.a(host_utils.cpp.o): in function `MPI::Intracomm::Create_graph(int, int const*, int const*, bool) const':
/opt/nvidia/hpc_sdk/Linux_x86_64/dev/comm_libs/mpi/include/openmpi/ompi/mpi/cxx/intracomm_inln.h:146: undefined reference to `MPI::Comm::Comm()'
/usr/bin/ld: libquda_test.a(host_utils.cpp.o):/opt/nvidia/hpc_sdk/Linux_x86_64/dev/comm_libs/mpi/include/openmpi/ompi/mpi/cxx/topology_inln.h:28: more undefined references to `MPI::Comm::Comm()' follow
/usr/bin/ld: libquda_test.a(host_utils.cpp.o):(.data+0x78): undefined reference to `MPI::Datatype::Free()'
/usr/bin/ld: libquda_test.a(host_utils.cpp.o):(.data+0x558): undefined reference to `MPI::Win::Free()'
```

Since `nvc++` is considered experimental for QUDA, and the work around is simple, I don't consider this a blocker.

So I think the only remaining issue from my perspective is resolution of the `-I` -> `-isystem` for Eigen.

File: lib/CMakeLists.txt
Comment (maddyscientist) [reply lib/CMakeLists.txt]: I just fixed this since it was quick 😉 a2305638971473b8ae24580c3d5ba809a9e63961

File: lib/CMakeLists.txt
Comment (mathiaswagner) [inline lib/CMakeLists.txt]: @cpviolator I'd like to revert this change as a private dependency should not be propagated to a downstream build. If you have a reproducer for the issue can you share the details?

**Comment by maddyscientist** [comment]
Some updates with respect to using `nvc++`
* The warnings I mentioned above when compiling QMP are now fixed in https://github.com/usqcd-software/qmp/pull/16.  Hopefully this will be merged before this PR so we can point CMake at an updated QMP hash that includes this fix.
* The issue with `Eigen` not being included with `-isystem` was a CMake bug.  The fix will be included in CMake 3.22, and I've verified this fixes the issue.

File: lib/CMakeLists.txt
Comment (cpviolator) [reply lib/CMakeLists.txt]: I've attached a reproducer.
[compile_eigen_repro.txt](https://github.com/lattice/quda/files/7471240/compile_eigen_repro.txt)
The script assumes:
```
Currently Loaded Modules:
  1) StdEnv (S)   2) cmake/3.20.2   3) cuda/11.4.1   4) gcc/8.3.1   5) spectrum-mpi/rolling-release
```
Everything is automatic except at the point where the revert is applied, then you have to interact to continue. The error I see is:
```
[100%] Linking CXX executable t_minvert
/usr/tce/packages/gcc/gcc-8.3.1/rh/usr/bin/../libexec/gcc/ppc64le-redhat-linux/8/ld: cannot find -lEigen
collect2: error: ld returned 1 exit status
make[2]: *** [mainprogs/tests/CMakeFiles/t_minvert.dir/build.make:116: mainprogs/tests/t_minvert] Error 1
make[1]: *** [CMakeFiles/Makefile2:782: mainprogs/tests/CMakeFiles/t_minvert.dir/all] Error 2
```
for each of the chroma executables.

**Comment by maddyscientist** [comment]
Pushed a commit that updates to use latest QMP (fixes warnings with nvc++).   While there's some outstanding warnings with nvc++ (when building QIO - requires merge of https://github.com/usqcd-software/qio/pull/14), and the link issue with nvc++ when building with QMP/QIO enabled (for which we have a work around), I don't see these as blockers for this PR.  So approving this PR.  Of course, @mathiaswagner if you happen to fix the link issue before we merge, I'm happy 😃

**Review by maddyscientist — APPROVED** [review]

File: lib/CMakeLists.txt
Comment (mathiaswagner) [reply lib/CMakeLists.txt]: Thanks. Fixed in 6736491

**Comment by mathiaswagner** [comment]
Any thoughts on removing `QUDA_PROPAGATE_CXX_FLAGS` ?

**Comment by bjoo** [comment]
Not yet... At some point I’d have to re-work the HIP section too to use the new HIP Language support.
I don’t know if it could come in useful there...
Best,
 B

--
Balint Joo,  Oak Ridge Leadership Computing Facility,
Oak Ridge National Laboratory
P.O. Box 2008, 1 Bethel Valley Road, Oak Ridge, TN 37831, USA
email: joob AT ornl.gov.  Tel: +1-757-912-0566 (cell, remote)

From: Mathias Wagner ***@***.***>
Reply-To: lattice/quda ***@***.***>
Date: Thursday, November 4, 2021 at 9:46 AM
To: lattice/quda ***@***.***>
Cc: Balint Joo ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [lattice/quda] rebasing moderner-cmake on gk (#1156)
Resent-From: ***@***.***>
Resent-Date: Thursday, November 4, 2021 at 9:46 AM


Any thoughts on removing QUDA_PROPAGATE_CXX_FLAGS ?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_1156-23issuecomment-2D960964331&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=gZTQdGtANNqLTYnsuOCox85Ts6JgD1VSQraWcyMGnKWDr_71OVT3UFumXv28q8sw&s=sKkWMfo7Qj8LQlPOQlMgoX6gAjEOU_Ttyr3dnzAdMWs&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAEPL2GP4YCOCG4WOSXUXSDUKKFC7ANCNFSM45WPGPXQ&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=gZTQdGtANNqLTYnsuOCox85Ts6JgD1VSQraWcyMGnKWDr_71OVT3UFumXv28q8sw&s=Z5Sfe9Bv1GDkLoSSZbDSaXoQtjxxBuGuKY2zHqeQYgA&e=>.
Triage notifications on the go with GitHub Mobile for iOS<https://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=gZTQdGtANNqLTYnsuOCox85Ts6JgD1VSQraWcyMGnKWDr_71OVT3UFumXv28q8sw&s=WxMi6o59INrP4Vo7lJkIqOCaT9QgLhh-zKCwrFKAXJs&e=> or Android<https://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26referrer-3Dutm-5Fcampaign-253Dnotification-2Demail-2526utm-5Fmedium-253Demail-2526utm-5Fsource-253Dgithub&d=DwMCaQ&c=CJqEzB1piLOyyvZjb8YUQw&r=WG01dcOd8ZFLFFNkrpbj6A&m=gZTQdGtANNqLTYnsuOCox85Ts6JgD1VSQraWcyMGnKWDr_71OVT3UFumXv28q8sw&s=2woaoH0fqd-Ol-LKpMm98BdB8OSwCKtyfEpTirXKOXE&e=>.

**Comment by mathiaswagner** [comment]
I'll leave it for another PR to remove the QUDA_PROPAGATE_CXX_FLAGS then.

**Pull request merged**
